### PR TITLE
[gcov] use __gcov_dump() instead for gcc 11

### PR DESCRIPTION
**Pull request by bukepo**
__gcov_flush is removed from gcc 11.

### CODE REVIEW

File: examples/apps/cli/main.c [hunk 1]
```diff
@@ -98,7 +98,7 @@ int main(int argc, char *argv[])
     {
         alarm(0);
 #if OPENTHREAD_ENABLE_COVERAGE
-        __gcov_flush();
```
Comment (simonlingoogle) [inline examples/apps/cli/main.c:101]: We need to call flash or dump according to the current compiler version?

**Review by simonlingoogle — COMMENTED** [review]
Thanks for trying to resolve this issue.

File: examples/apps/cli/main.c (hunk 1 above)
Comment (bukepo) [reply examples/apps/cli/main.c:101]: Yes, I use this PR to check if the __gcov_dump is available in all versions.

File: src/lib/platform/reset_util.h [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+#define flush_gcov __gcov_flush
+#endif // __GNUC__ >= 11
+#else
+#define flush_gcov()
+#endif // defined(OPENTHREAD_ENABLE_COVERAGE) && OPENTHREAD_ENABLE_COVERAGE && defined(__GNUC__)
+
+#if defined(__linux__) || defined(__APPLE__)
+#include <setjmp.h>
+#include <unistd.h>
+jmp_buf gResetJump;
```
Comment (simonlingoogle) [inline src/lib/platform/reset_util.h:53]: every source file that includes reset_util.h would have a new instance of `jmp_buf gResetJump`. should put the definition in a source file?

```suggestion
extern jmp_buf gResetJump;
```

File: src/lib/platform/reset_util.h [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef OT_EXAMPLES_PLATFORMS_UTILS_RESET_UTIL_H_
+#define OT_EXAMPLES_PLATFORMS_UTILS_RESET_UTIL_H_
```
Comment (simonlingoogle) [inline src/lib/platform/reset_util.h:30]: why lib/platform rather than src/platform/posix ?

File: src/lib/platform/reset_util.h (hunk 2 above)
Comment (bukepo) [reply src/lib/platform/reset_util.h:53]: Intended to avoid this file by used multiple times.

File: src/lib/platform/reset_util.h (hunk 3 above)
Comment (bukepo) [reply src/lib/platform/reset_util.h:30]: also used by simulation platform.

Comment (simonlingoogle) [reply src/lib/platform/reset_util.h]: make sense. Thanks.

File: src/lib/platform/reset_util.h (hunk 2 above)
Comment (simonlingoogle) [reply src/lib/platform/reset_util.h:53]: I guess there will be a duplicate symbols linking issue if included multiple times?

Comment (bukepo) [reply src/lib/platform/reset_util.h:53]: yes, unless `--allow-multiple-definition` is explicitly used. This file is supposed to be included once in a single object file.

Comment (simonlingoogle) [reply src/lib/platform/reset_util.h:53]: I think it's fine as long as the misuse can be detected.

File: src/lib/platform/reset_util.h [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+#else
+#define flush_gcov()
+#endif // defined(OPENTHREAD_ENABLE_COVERAGE) && OPENTHREAD_ENABLE_COVERAGE && defined(__GNUC__)
+
+#if defined(__linux__) || defined(__APPLE__)
+#include <setjmp.h>
+#include <unistd.h>
+jmp_buf gResetJump;
+
+#define OT_SETUP_RESET(ARGV)   \
```
Comment (simonlingoogle) [inline src/lib/platform/reset_util.h:55]:
```suggestion
#define OT_SETUP_RESET_JMP(ARGV)   \
```

It's only necessary when we choose `longjmp` as the reset strategy. 
So, better make it explicit that it's for `longjmp` ?

**Review by simonlingoogle — APPROVED** [review]
LGTM 👍 

with nits

**Review by abtink — APPROVED** [review]
👍 Thanks

**Review by jwhui — APPROVED** [review]

**Pull request merged**